packages/components/src/tools-panel/tools-panel/hook.ts

```diff
@@ -81,44 +97,55 @@ export function useToolsPanel( props ) {
 	};
 
 	// Manage and share display state of menu items representing child controls.
-	const [ menuItems, setMenuItems ] = useState( {} );
+	const [ menuItems, setMenuItems ] = useState< ToolsPanelMenuItems >( {
+		default: {},
+		optional: {},
+	} );
```

Comment:
> Is it necessary to initialise menuItems to this default value, instead of leaving it undefined ?

This seemed to be the most straightforward means of addressing the typing errors reported. It initially began as an empty object. Switching to `undefined` resulted in additional errors. Later, when attempting to resolve the typing errors around dynamically updating the individual items within `menuItems` (via the `menuGroup`) we found we needed to add `default` and `optional` as mandatory properties on the `ToolsPanelMenuItems` type. This forced the initialization of the state to include those properties.

An [alternative approach](https://github.com/WordPress/gutenberg/pull/34981) has been worked up separately to this PR. For the removal of the default object when declaring the `menuItems` state, we're reducing the readability of 3 or 4 other places in the code. Personally, I like the simpler, clearer, approach in place currently.

I'm definitely happy to change that if you have a better solution though.

> Also a nit: do you think it would be better, for readability's sake, to move this state declaration before flagItemCustomization (which calls setMenuItems internally) ?

✅  Done.

Reply:
Thank you for showcasing the alternative approach in a [separate PR](https://github.com/WordPress/gutenberg/pull/34981#pullrequestreview-759529019). I actually agree with you, let's keep the default value of the object as-is :)